remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,768 @@
+      for (auto &block : func) {
+        for (auto &inst : block) {
+          fav.AddInstruction(&inst);
+        }
+      }
+      // if the analysis succeeds for this function, add the AAMDNodes
+      if (fav.Analyze()) {
+        LOG(INFO) << "Offsets: " << fav.offset_map.size();
+        LOG(INFO) << "Aliases: " << fav.alias_map.size();
+        LOG(INFO) << "Excluded: " << fav.exclude.size();
+        AddAAMDNodes(fav.alias_map, aamd_info.slot_aamds);
+        // Perform live set analysis
+        LiveSetBlockVisitor LSBV(func, slots, aamd_info.slot_scopes, fav.offset_map, bb_func->getFunctionType(), &dl);
+        LSBV.Visit();
+        // repeat, but now ready to remove
+        LSBV.RemoveDeadStores();
+      }
+    }
+  }
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(llvm::Function &func_,
+    const std::vector<StateSlot> &state_slots_,
+    const ScopeMap &scope_to_offset_,
+    const ValueToOffset &val_to_offset_,
+    const llvm::FunctionType *lifted_func_ty_,
+    const llvm::DataLayout *dl_)
+  : func(func_),
+    val_to_offset(val_to_offset_),
+    scope_to_offset(scope_to_offset_),
+    state_slots(state_slots_),
+    curr_wl(),
+    next_wl(),
+    block_map(),
+    to_remove(),
+    lifted_func_ty(lifted_func_ty_),
+    func_used(),
+    on_remove_pass(false),
+    dl(dl_) {
+      for (auto &block_it : func) {
+        auto block = &block_it;
+        auto succ_block_it = successors(block);
+        if (succ_block_it.begin() == succ_block_it.end()) {
+          curr_wl.push_back(block);
+        }
+      }
+    }
+
+// Visit the basic blocks in the worklist and update the block_map.
+void LiveSetBlockVisitor::Visit(void) {
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    for (auto block : curr_wl) {
+      if (VisitBlock(block)) {
+        // Updates have been made
+        // Add predecessors to next rotation
+        for (llvm::BasicBlock *pred : predecessors(block)) {
+          next_wl.push_back(pred);
+        }
+      }
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+}
+
+// Return true if the call accesses the state struct.
+// If a single slot is accessed, set the slotn to that slot.
+bool LiveSetBlockVisitor::CallAccessesState(llvm::CallInst *call, uint64_t *slotn) {
+  slotn = nullptr;
+  for (auto &arg_it : call->arg_operands()) {
+    auto arg = arg_it->stripPointerCasts();
+    if (val_to_offset.count(arg)) {
+      auto offset = val_to_offset.at(arg);
+      if (offset != 0) {
+        // if we access a single non-zero offset, mark just that offset
+        *slotn = state_slots[offset].i;
+      }
+      return true;
```

Comment:
This is premature; what if the first argument was for slot 1, and the second for slot 0? 

What we really want is so pass in a reference to the live set. Revive any slots that are seen, but if we see slot offset 0, revive everything. Then the function doesn't need to return `true` or `false` because it will have done the work.